apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,524 @@
+			QuantizedValue: account.LatestValue.QuantizedValue.BigInt(),
+			TimestampNs:    account.LatestValue.TimestampNs,
+		}
+	}
+
+	return polledVals, nil
+}
+
+func (sci *SolanaContractInteracter) BatchPushToContract(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) error {
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(priceUpdates))
+	sigChan := make(chan solana.Signature, len(priceUpdates))
+	priceUpdatesBatches := sci.batchPriceUpdates(priceUpdates)
+	for _, priceUpdateBatch := range priceUpdatesBatches {
+		wg.Add(1)
+		go func(priceUpdateBatch map[InternalEncodedAssetId]AggregatedSignedPrice) {
+			defer wg.Done()
+			err := sci.limiter.Wait(context.Background())
+			if err != nil {
+				errChan <- fmt.Errorf("rate limiter error: %w", err)
+				return
+			}
+			sig, err := sci.pushLimitedBatchUpdateToContract(priceUpdateBatch)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to push batch: %w", err)
+			} else {
+				sigChan <- sig
+			}
+		}(priceUpdateBatch)
+	}
+
+	wg.Wait()
+	close(errChan)
+	close(sigChan)
+	// Collect any errors
+	var errs []error
+	for err := range errChan {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("batch push encountered %d errors: %v", len(errs), errs)
+	}
+
+	sigs := []string{}
+	for sig := range sigChan {
+		sigs = append(sigs, sig.String())
+	}
+
+	sci.logger.Info().
+		Int("numUpdates", len(priceUpdates)).
+		Strs("batchTransactionSigs", sigs).
+		Msg("Successfully pushed batch updates to contract")
+
+	return nil
+}
+
+func (sci *SolanaContractInteracter) batchPriceUpdates(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) []map[InternalEncodedAssetId]AggregatedSignedPrice {
+	priceUpdatesBatches := []map[InternalEncodedAssetId]AggregatedSignedPrice{}
+
+	priceUpdatesBatch := make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+	i := 0
+	for encodedAssetId, priceUpdate := range priceUpdates {
+		priceUpdatesBatch[encodedAssetId] = priceUpdate
+		i++
+		if len(priceUpdatesBatch) == sci.batchSize || i == len(priceUpdates) {
+			batchCopy := make(map[InternalEncodedAssetId]AggregatedSignedPrice, len(priceUpdatesBatch))
+			for k, v := range priceUpdatesBatch {
+				batchCopy[k] = v
+			}
+			priceUpdatesBatches = append(priceUpdatesBatches, batchCopy)
+			priceUpdatesBatch = make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+		}
+	}
+
+	return priceUpdatesBatches
+}
+
+func (sci *SolanaContractInteracter) pushLimitedBatchUpdateToContract(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) (solana.Signature, error) {
+
+	if len(priceUpdates) > MAX_BATCH_SIZE {
+		return solana.Signature{}, fmt.Errorf("Batch size exceeds limit, skipping update")
+	}
+
+	randomId, err := rand.Int(rand.Reader, big.NewInt(256))
+	if err != nil {
+		return solana.Signature{}, fmt.Errorf("failed to generate random ID: %w", err)
+	}
+	treasuryId := uint8(randomId.Uint64())
+
+	treasuryAccount := sci.treasuryAccounts[treasuryId]
```

Comment:
this pattern of using a pool of treasury accounts and picking one randomly for each push is for security reasons?

Reply:
It is for write throughput reasons. If you have multiple clients pushing and they're all trying to interact with the same treasury account it could create write concurrency issues.